blog/2023-02-02-good-cognitive-friction.md

```diff
@@ -0,0 +1,251 @@
+will be written with just `var` and mountains of linters will be written to protect you from
+shooting yourself in the foot.
+
+Going back to this concept of "good cognitive friction". If you need to type a few more characters in
+order to make a variable reassignable (`let var` versus `let`), you will likely just use `let` most
+of the time, and the world will be a better place with less bugs and happier developers.
+
+## The inflight Connection
+
+So how is all this related to cloud development?
+
+One of the very cool things about immutable state is that the compiler can create as many copies of
+it as needed. If the compiler has a guarantee that a blob of data will never change over the
+lifetime (and space) of the system, it can simply distribute it where it is needed.
+
+This means, that in Wing, immutable data can be seamlessly referenced from any [inflight
+context](https://docs.winglang.io/concepts/inflights).
+
+Let's look at a very simple example just to explain the idea:
+
+```js
+bring cloud;
+
+let my_array = ["hello", "world"];
+
+new cloud.Function(inflight (_: str) => { 
+  assert(my_array.length == 2); 
+}) as "test";
+```
+
+So what's going on here? We have defined a cloud function that simply references `my_array`. As much
+as this looks simple and intuitive, the compiler actually had to do a bit of work to make this
+happen. As a reminder, a `cloud.Function` represents a cloud compute platform (such as AWS Lambda).
+This means that the code inside the `inflight` block is going to be executed sometime in the future,
+on some other machine. Completely isolated from the original memory space in which `my_array` was
+defined.
+
+Since our array is immutable, the compiler can safely clone it and bundle it together with the code
+that runs inside the cloud function.
+
+> In the future, the compiler will be able to identify that `my_array.length` itself is immutable,
+> and will only copy its value (see [#1251](https://github.com/winglang/wing/issues/1251)).
+
+If we try to reference a reassignable variable from inflight code:
+
+```js
+let var s = "hello";
+
+new cloud.Function(inflight (_: str) => {
+  print(s);
+  //    ^ Cannot capture a reassignable variable "s"
+});
+```
+
+If we try to reference a mutable collection from inflight code:
+
+```js
+bring cloud;
+
+let my_array = MutArray<num>[1,2,3,4];
+
+new cloud.Function(inflight (_: str) => {
+  assert(my_array.length == 4);
+  //     ^^^^^^^^ Cannot reference 'my_array' of type 'MutArray<num>' from an inflight context
+});
+```
+
+In this case as well, the compiler won't allow us to reference a mutable object within an inflight
+context, because it won't be able to guarantee correctness.
+
+Unsupported yet, but we will also have `freeze()` to cover you in case you want to reference a
+snapshot of a mutable collection:
+
+```js
+let mut_arr = MutArray<num>[1,2,3];
+let arr = mut_arr.freeze();
```

Comment:
I think `freeze` makes it sound like we are modifying `mut_arr`, when we want to convey we are creating a copy IIUC. Though `to_immut` doesn't roll off the tongue to me personally so maybe we keep this for now...

Reply:
OK, changed to `clone()` (and `clone_mut()`).